### PR TITLE
Teach rectangle selection to invert previous selection 

### PR DESCRIPTION
**Pull request by Don-Vito**
Stacked upon https://github.com/files-community/Files/pull/2716.
(that is why the change looks bigger than it is :blush:)

In the previous PR when CTRL is clicked, rectangle selection preserves
previous selection.

Here we extend the logic to invert selection on items that were
previously selected, like in Windows Explorer:

InvertSelection

### CODE REVIEW

**Comment by Don-Vito** [comment]
Resolving conflicts.

File: Files/UserControls/Selection/ISelectedItems.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Files.UserControls.Selection
+{
+    public interface ISelectedItems
```
Comment (gave92) [inline Files/UserControls/Selection/ISelectedItems.cs:3]: Is this interface necessary? It's identical to `IList`

File: Files/UserControls/Selection/GenericSelectedItems.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+﻿using System.Collections.Generic;
+
+namespace Files.UserControls.Selection
+{
+    public class GenericSelectedItems : ISelectedItems
```
Comment (gave92) [inline Files/UserControls/Selection/GenericSelectedItems.cs:5]: Same comment as ISelectedItems

File: Files/UserControls/Selection/RawSelectedItems.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+﻿using System.Collections;
+
+namespace Files.UserControls.Selection
+{
+    public class RawSelectedItems : ISelectedItems
```
Comment (gave92) [inline Files/UserControls/Selection/RawSelectedItems.cs:5]: What's the difference with GenericSelectedItems?

File: Files/UserControls/Selection/ISelectedItems.cs (hunk 1 above)
Comment (Don-Vito) [reply Files/UserControls/Selection/ISelectedItems.cs:3]: @gave92 - I know it looks crazy :smile:

I didn't touch C# for a while, but I wanted to apply the same logic to both `Collections.IList` and `Collections.Generics.IList<object>`.

The reason for this is that I wanted to reuse the same strategy for both list and grid, but their respective Selected items have different types.

File: Files/UserControls/Selection/GenericSelectedItems.cs (hunk 2 above)
Comment (Don-Vito) [reply Files/UserControls/Selection/GenericSelectedItems.cs:5]: If there is a better solution from the problem above I will happily get rid of this :blush:

File: Files/UserControls/Selection/RawSelectedItems.cs (hunk 3 above)
Comment (Don-Vito) [reply Files/UserControls/Selection/RawSelectedItems.cs:5]: and of this :blush:

File: Files/UserControls/Selection/ISelectedItems.cs (hunk 1 above)
Comment (gave92) [reply Files/UserControls/Selection/ISelectedItems.cs:3]: Suggestion: `uiElement.SelectedItems.Cast<object>().ToList()` will convert an `IList` to an `IList<object>`
The reverse should be possible as well

Comment (Don-Vito) [reply Files/UserControls/Selection/ISelectedItems.cs:3]: @gave92  - I cannot do it as I need a side-effect on the original collection

Comment (gave92) [reply Files/UserControls/Selection/ISelectedItems.cs:3]: And the reverse? var generic = (System.Collections.IList)uiElement.SelectedItems;

Comment (Don-Vito) [reply Files/UserControls/Selection/ISelectedItems.cs:3]: Crashes the COM object :smile:

**Comment by gave92** [comment]
@Don-Vito you should find a PR on your repo with an slightly different approach, let me know if you agree with it.
Also great job with this, everything else seems to be working good 👍

**Comment by Don-Vito** [comment]
> 
> 
> @Don-Vito you should find a PR on your repo with an slightly different approach, let me know if you agree with it.
> Also great job with this, everything else seems to be working good 👍

@gave92  - Yeah, considered this as well, but was concerned with implementing a very yide interface.
(thus invented one of my own).

WDYT of implementing ICollection rather than IList - it should be sufficient for the strategy and it somewhat smaller
(I can do it not to waste your time)? 

Otherwise let's merge yours :blush:

**Comment by gave92** [comment]
Yup that's even better, if you can do that, then we can merge this. Thank you

**Review by gave92 — APPROVED** [review]

**Comment by Don-Vito** [comment]
@gave92 - sorry for the delay.. unrelated to this PR noticed that shift selection sometimes clashes with ctrl selection (when you move and click very fast). I will open a separate PR for fixing this.

**Pull request merged**